### ISSUE TITLE
Discussion: DIP check should distinguish data types from services

### ISSUE THREAD

**Issue by kgatilin** [mshogin/archlint#100]
## Problem

The DIP rule flags every module that has structs without traits. This produces false positives on data-only modules: domain value objects, DTOs, config structs.

Example from deskd — 11 DIP violations, but only 2 are real architectural problems. The other 9 are modules that hold pure data (no I/O, no behavior, no methods beyond getters).

## The distinction

**Service** (struct with behavior — needs a trait):
```rust
// 5 methods, does I/O, manages subprocess — REAL DIP violation
pub struct AgentProcess { ... }
impl AgentProcess {
    pub fn start(&mut self) -> Result<()> { /* spawns process */ }
    pub fn send_task(&mut self, task: &str) -> Result<TurnResult> { /* I/O */ }
    pub fn stop(&mut self) -> Result<()> { /* kills process */ }
}
```

**Data type** (struct with fields — trait is meaningless):
```rust
// Pure data, no behavior — NOT a DIP violation
pub struct Message {
    pub id: String,
    pub source: String,
    pub target: String,
    pub payload: serde_json::Value,
}
```

## Current behavior

Both get flagged as "module has N structs but no trait definitions; consider introducing traits."

## Discussion points

1. **Should archlint distinguish these cases?** Possible heuristics:
   - Struct has `impl` block with methods that take `&mut self` or do I/O → service
   - Struct has only `pub` fields, `derive` macros, no `impl` → data type
   - Struct methods are pure (`&self` only, no side effects) → data type

2. **Or is the current approach correct?** Maybe ALL structs should have traits, even data types, and the "false positive" argument is just developer convenience?

3. **Middle ground?** Different severity levels:
   - Service without trait → `personal` (fix it)
   - Data type without trait → `telemetry` (informational)

4. **Cross-module trait detection?** In deskd, domain value objects (Message, Task) DO have corresponding traits — just in a different module (ports/bus.rs, ports/store.rs). The linter can't see this association. Should it follow the dependency graph to check if a type is referenced in a trait signature elsewhere?

## Real-world data from deskd

| Module | Structs | Has methods with I/O? | Real DIP? |
|--------|---------|----------------------|-----------|
| domain/message.rs | 3 | No | No — MessageBus trait in ports/ |
| domain/task.rs | 3 | No | No — TaskRepository trait in ports/ |
| domain/statemachine.rs | 4 | No | No — StateMachineRepository in ports/ |
| domain/context.rs | 5 | No | Partial — missing ContextRepository |
| config.rs | 14 | load() only | No — config parsing |
| infra/dto.rs | 15 | No | No — serde ACL layer |
| bin/session2graph.rs | 13 | No | No — standalone binary |
| app/agent.rs | 6 | Yes (subprocess I/O) | **Yes** — should impl Executor trait |
| app/acp.rs | 4 | Yes (subprocess I/O) | **Yes** — should impl Executor trait |
| app/mcp.rs | 3 | Mixed | Partial — code org issue |
| app/graph.rs | 6 | Yes (subprocess) | Low priority |

Would love to hear thoughts on the right approach here.
